Confusing error message "this doc comment doesn't document anything" when there is a syntax error right after a comment

I tried this code:

```rust
/// Some documentation. Below "async" is before "pub" by mistake.
async pub fn hello() {}

fn main() {
}
```

I expected to see this happen: a clear error message telling me about the mistake (`async` before `pub`).

Instead, this happened:

```
error: expected item after doc comment
 --> src/main.rs:1:1
  |
1 | /// Some documentation. Below "async" is before "pub" by mistake.
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ this doc comment doesn't document anything
```

This is a confusing error message. Without the comment, the error is clearer (though arguably still not completely obvious):

```
error: expected item, found keyword `async`
 --> src/main.rs:1:1
  |
1 | async pub fn hello() {}
  | ^^^^^ expected item
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.43.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.43.0
LLVM version: 10.0
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
No backtrace.
```

</p>
</details>

Agreed, I just hit this as well. It also seems to be inconsistent based on the syntax error. For example, a missing identifier syntax error outputs a helpful error
```rust
/// hello world
async fn {}
```
```
error: expected identifier, found `{`
 --> src/lib.rs:2:10
  |
2 | async fn {}
  |          ^ expected identifier
```
while a different syntax error prints out a different error than the first one:
```rust
/// hello world
<>
```
```
error[E0585]: found a documentation comment that doesn't document anything
 --> src/lib.rs:1:1
  |
1 | /// hello world
  | ^^^^^^^^^^^^^^^
  |
  = help: doc comments must come before what they document, maybe a comment was intended with `//`?
```

The following code snippet can trigger this bug as well:
```rust
enum TestEnum {
    Works,
    /// Some documentation
    Self,
}
```
Current output:
```
error[E0585]: found a documentation comment that doesn't document anything
 --> test_enum.rs:3:5
  |
1 | enum TestEnum {
  |      -------- while parsing this enum
2 |     Works,
3 |     /// Some documentation
  |     ^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: doc comments must come before what they document, if a comment was intended use `//`
  = help: enum variants can be `Variant`, `Variant = <integer>`, `Variant(Type, ..., TypeN)` or `Variant { fields: Types }`

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0585`.
```

Expected output:
```
error: expected identifier, found keyword `Self`
 --> test_enum.rs:4:5
  |
1 | enum TestEnum {
  |      -------- while parsing this enum
...
4 |     Self,
  |     ^^^^ expected identifier, found keyword
  |
  = help: enum variants can be `Variant`, `Variant = <integer>`, `Variant(Type, ..., TypeN)` or `Variant { fields: Types }`

error: aborting due to 1 previous error
```

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -301,13 +301,6 @@ impl<'a> Parser<'a> {
         &mut self,
         recover: bool,
     ) -> PResult<'a, (Ident, IdentIsRaw)> {
-        if let TokenKind::DocComment(..) = self.prev_token.kind {
-            return Err(self.dcx().create_err(DocCommentDoesNotDocumentAnything {
-                span: self.prev_token.span,
-                missing_comma: None,
-            }));
-        }
-
         let valid_follow = &[
             TokenKind::Eq,
             TokenKind::Colon,
@@ -319,6 +312,15 @@ impl<'a> Parser<'a> {
             TokenKind::CloseDelim(Delimiter::Brace),
             TokenKind::CloseDelim(Delimiter::Parenthesis),
         ];
+        if let TokenKind::DocComment(..) = self.prev_token.kind
+            && valid_follow.contains(&self.token.kind)
+        {
+            let err = self.dcx().create_err(DocCommentDoesNotDocumentAnything {
+                span: self.prev_token.span,
+                missing_comma: None,
+            });
+            return Err(err);
+        }
 
         let mut recovered_ident = None;
         // we take this here so that the correct original token is retained in

--- a/tests/ui/parser/doc-before-bad-variant.rs
+++ b/tests/ui/parser/doc-before-bad-variant.rs
@@ -1,0 +1,6 @@
+enum TestEnum {
+    Works,
+    /// Some documentation
+    Self, //~ ERROR expected identifier, found keyword `Self`
+    //~^ HELP enum variants can be
+}

--- a/tests/ui/parser/doc-before-bad-variant.stderr
+++ b/tests/ui/parser/doc-before-bad-variant.stderr
@@ -1,0 +1,13 @@
+error: expected identifier, found keyword `Self`
+  --> $DIR/doc-before-bad-variant.rs:4:5
+   |
+LL | enum TestEnum {
+   |      -------- while parsing this enum
+...
+LL |     Self,
+   |     ^^^^ expected identifier, found keyword
+   |
+   = help: enum variants can be `Variant`, `Variant = <integer>`, `Variant(Type, ..., TypeN)` or `Variant { fields: Types }`
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/parser/doc-before-syntax-error.rs
+++ b/tests/ui/parser/doc-before-syntax-error.rs
@@ -1,0 +1,2 @@
+/// Some documentation
+<> //~ ERROR expected identifier

--- a/tests/ui/parser/doc-before-syntax-error.stderr
+++ b/tests/ui/parser/doc-before-syntax-error.stderr
@@ -1,0 +1,8 @@
+error: expected identifier, found `<`
+  --> $DIR/doc-before-syntax-error.rs:2:1
+   |
+LL | <>
+   | ^ expected identifier
+
+error: aborting due to 1 previous error
+

